File: backend/src/main/scala/domains/problemset/http/api/AddProblemToProblemSet.scala
package domains.problemset.http.api

import domains.problemset.http.mapper.ProblemSetHttpResponseMappers
import domains.problemset.http.mapper.ProblemSetHttpRequestMappers



import domains.problemset.http.*
import domains.problemset.http.codec.ProblemSetHttpCodecs.given
import cats.effect.IO
import domains.problemset.application.ProblemSetCommands
import domains.problemset.objects.request.{AddProblemToProblemSetRequest}
import domains.problemset.objects.ProblemSetSlug
import org.http4s.HttpRoutes
import org.http4s.circe.CirceEntityCodec.*
import org.http4s.dsl.Http4sDsl
import org.http4s.dsl.io.*

object AddProblemToProblemSet:

  def routes(handlers: domains.auth.http.AuthenticatedHttpExecutor)(using Http4sDsl[IO]): HttpRoutes[IO] =
    HttpRoutes.of[IO] {
      case request @ POST -> Root / "api" / "problem-sets" / problemSetSlug / "problems" =>
        ProblemSetHttpRequestMappers.problemSetSlug(problemSetSlug) match
          case Left(message) =>
            ProblemSetHttpResponseMappers.validationErrorResponse(message)
          case Right(parsedProblemSetSlug) =>
            handlers.executeDecoded[
              AddProblemToProblemSetRequest,
              (ProblemSetSlug, AddProblemToProblemSetRequest),
              ProblemSetCommands.AddProblemResult
            ](
              request,
              ProblemSetHttpPlanDefinitions.addProblem
            )(addRequest => (parsedProblemSetSlug, addRequest))
    }

File: frontend/src/apis/problemset/AddProblemToProblemSet.ts
import type { APIWithSessionMessage } from '@/system/api/api-message'
import type { AddProblemToProblemSetRequest } from '@/objects/problemset/request/AddProblemToProblemSetRequest'
import type { ProblemSetDetail } from '@/objects/problemset/response/ProblemSetDetail'
import type { ProblemSetSlug } from '@/objects/problemset/ProblemSetSlug'
import { problemSetSlugValue } from '@/objects/problemset/ProblemSetSlug'

/** 向题集添加题目；输入题集 slug 和题目 slug 请求，输出更新后的题集详情。 */
export class AddProblemToProblemSet implements APIWithSessionMessage<ProblemSetDetail> {
  declare readonly responseType?: ProblemSetDetail
  readonly method = 'POST'
  readonly apiPath: string
  private readonly request: AddProblemToProblemSetRequest

  constructor(problemSetSlug: ProblemSetSlug, request: AddProblemToProblemSetRequest) {
    this.apiPath = `problem-sets/${problemSetSlugValue(problemSetSlug)}/problems`
    this.request = request
  }

  body(): AddProblemToProblemSetRequest {
    return this.request
  }
}
